[Editor] Stop drawing a shape outside of the page

Hydraulicus opened the pull request:
Issue: drawing a shape may have an unwanted path element added.
Steps to reproduce:
1. Go to https://mozilla.github.io/pdf.js/web/viewer.html;
2. Select the pen tool and draw a shape on the second page;
3. While drawing the shape, move the cursor out of the left-hand side of the second page, then stop the action.
On the third step, the shape will be extended with a path element that starts where the cursor left the page and ends at a random position on the page. Correct this issue to ensure that the path element is not added.

Fix: Incorrect End Position of Curve on Target Page
The issue of the curve's unintended end position is caused by obtaining improper coordinates from the end-drawing event. These coordinates are derived from an element other than the intended target page.

Changes:
To resolve this, I added a check to verify the element where the drawing ends. If the drawing ends on an element other than the target page, the coordinates registration are skipped.

Hydraulicus committed:
fixing bug of stop drawing shape outside page

Hydraulicus committed:
[Editor] Add validation for the target element of curve endpoints

This patch fixes a bug that caused incorrect curve shapes when an endpoint lies beyond the page boundaries. It adds a check for the endpoint's position, and if it is outside the page, the point is excluded from the shape's coordinates.

Hydraulicus committed:
clearing up

Snuffleupagus reviewed (commented):
- Please squash the commits, since we don't use separate fixup commits in this project.
 - Please write a better *commit message*, since a single line is rarely enough context (i.e. it must be possible to understand what a patch does and why on the Git command line as well).
 - Please add an integration-test for the changes, such that it won't regress in the future.
   Also, does the patch pass all existing tests?

Snuffleupagus commented on src/display/editor/draw.js:
```diff
@@ -834,7 +834,7 @@ class DrawingEditor extends AnnotationEditor {
     this._cleanup(false);
 
-    if (event) {
+    if (event && event.target === parent.div) {
```
```suggestion
    if (event?.target === parent.div) {
```

Hydraulicus committed:
[Editor] Add validation for the target element of curve endpoints

This patch fixes a bug that caused incorrect curve shapes when an endpoint lies beyond the page boundaries. It adds a check for the endpoint's position, and if it is outside the page, the point is excluded from the shape's coordinates.

Hydraulicus committed:
Merge remote-tracking branch 'origin/fix-pen-outside-page' into fix-pen-outside-page

Snuffleupagus commented on test/integration/ink_editor_spec.mjs:
```diff
@@ -1011,6 +1012,60 @@ describe("Ink Editor", () => {
       );
     });
   });
+  describe("The pen-drawn shape must maintain correct curvature regardless of the page it is drawn on or whether the curve's endpoint lies within or beyond the page boundaries.", () => {
```
```suggestion
  describe("The pen-drawn shape must maintain correct curvature regardless of the page it is drawn on or whether the curve's endpoint lies within or beyond the page boundaries", () => {
```

Hydraulicus commented:
> * Please squash the commits, since we don't use separate fixup commits in this project.
> * Please write a better _commit message_, since a single line is rarely enough context (i.e. it must be possible to understand what a patch does and why on the Git command line as well).
> * Please add an integration-test for the changes, such that it won't regress in the future.
>   Also, does the patch pass all existing tests?

* I have squashed my commits and synchronised with the latest master. Kindly check. 
* I have added details into commit description based on description of previous commits. Please review.
* I added integration test for this case. Please check. 
Yes the patch pass existing tests. I see, al least in my environment, some test are failed, and this patch don't add another failed tests.

Snuffleupagus commented on test/integration/test_utils.mjs:
```diff
@@ -835,6 +835,16 @@ async function cleanupEditing(pages, switcher) {
   }
 }
 
+function normalizeSpaces(d) {
```
Please use a slightly more descriptive function name, since "normalizeSpaces" sounds like something that'd be used when parsing regular text.

Snuffleupagus reviewed (commented):
The patch now consists of no less than five commits, please squash them together; see also https://github.com/mozilla/pdf.js/wiki/Squashing-Commits

calixteman commented on test/integration/test_utils.mjs:
```diff
@@ -835,6 +835,16 @@ async function cleanupEditing(pages, switcher) {
   }
 }
 
+function normalizeSpaces(d) {
```
nit: since this function has only one use, you can inline it.
That said, do you have an example of a string having those extra spaces ?

Hydraulicus replied:
Here we see the same path generated on 1st and 2nd pages. And difference is in redundant whitespaces, that don't allows to make correct comparisation. Thus I just clearing them out

Hydraulicus replied:
I will make refactoring of my code according to your comments. Thanks

Hydraulicus replied:
You notes a good point. In my current case with simple curves the d attributes looks the same in both pages.  Thus no need to make this normalisation. So probably currently I will not pollute the codebase and will not add this normalisation, and will remove the function.  Are you don't mind?

calixteman replied:
Yeah there's no need to add something useless.
That said, those extra spaces are likely coming from:
https://github.com/mozilla/pdf.js/blob/877f69886c3e000fb7be2d82d536476032f2b8e4/src/display/editor/drawers/inkdraw.js#L204

Hydraulicus commented:
@Snuffleupagus kindly review https://github.com/mozilla/pdf.js/pull/19373 
It is instead of this PR.

Snuffleupagus commented:
> kindly review #19373. Here is fulfilled all comments, reqirements and is made squash.
> It is instead of this PR.

*For future reference:* Please fix the commit history in an existing PR, rather than re-submitting a new PR since that make it more difficult to follow the discussion and ensure that previous review comments have been addressed.
